Spring 2021 README and docs updates (#331)
UMM-CSci-3601/3601-iteration-template: README.md

> [![BCH compliance](https://bettercodehub.com/edge/badge/UMM-CSci-3601/3601-iteration-template?branch=master)](https://bettercodehub.com/)

Author: ```suggestion
[![BCH compliance](https://bettercodehub.com/edge/badge/UMM-CSci-3601/3601-iteration-template?branch=main)](https://bettercodehub.com/)
```

previous labs. These are almost certainly not relevant to your project
and should be removed once you've gotten going.

## Setup

As in the previous labs, you'll be using VS Code and GitKraken. Once you've all joined your
group using GitHub classroom, you can clone your repository using the command line or GitKraken:

1. From the file menu, choose **Clone Repo**
2. Choose GitHub.com in the middle column (as the source location of your repo)
3. Browse to the location you'd like to put the local copy of this project repo
4. Select the correct repo from the list of repositories
5. Select **Clone the repo!**

### Make sure you have Mongo running on your computer

For all of this to work, it's critical that you have Mongo installed
and working, as described in the system setup documentation from the
beginning of the semester. If you're unsure if it's set up and
working correctly, try running `mongo`.

> If your MongoDB server isn't installed you'll likely get an error

Author: ```suggestion
If your MongoDB server isn't running you'll likely get an error
```

Should we maybe suggest what to do when it isn't running?

previous labs. These are almost certainly not relevant to your project
and should be removed once you've gotten going.

## Setup

As in the previous labs, you'll be using VS Code and GitKraken. Once you've all joined your
group using GitHub classroom, you can clone your repository using the command line or GitKraken:

1. From the file menu, choose **Clone Repo**
2. Choose GitHub.com in the middle column (as the source location of your repo)
3. Browse to the location you'd like to put the local copy of this project repo
4. Select the correct repo from the list of repositories
5. Select **Clone the repo!**

### Make sure you have Mongo running on your computer

For all of this to work, it's critical that you have Mongo installed
and working, as described in the system setup documentation from the
beginning of the semester. If you're unsure if it's set up and

> working correctly, try:
> 
> ```bash
> mongo
> ```

Author: ```suggestion
working correctly, try running `mongo`.
```

If your MongoDB server isn't installed you'll likely get an error
message like:

```text
Error: couldn't connect to server 127.0.0.1:27017, connection attempt failed: SocketException: Error connecting to 127.0.0.1:27017 :: caused by :: Connection refused :
```

> (Use CTRL-D to exit out of the `mongo` tool.)

Author: ```suggestion
(Type `exit` to exit out of the `mongo` tool.)
```

### Open the project in VS Code

Launch Visual Studio Code, and then choose `File -> Open Folder…`. Navigate to your clone
of the repo and choose `Open`.

You may see a dialog that looks like this if you don't already have the recommended extensions:

![Dialog suggesting installation of recommended extensions](https://user-images.githubusercontent.com/1300395/72710961-bf767500-3b2d-11ea-8ea4-fbbd39c78da5.png)

Don't worry if you don't get the dialog, it is probably because you already have them all installed.

Like in previous labs, click "Install All" to automatically install them.

### Installing the client dependencies

Before you start working you will need to install the dependencies for the client.

1. Move into the `client` directory (`cd client`)
2. Run `npm install`

### Enable ESLint in VS Code

> ESLint is a tool for checking the quality and style of your TypeScript
> and JavaScript code, and can provide valuable "live" feedback on
> your coding in VS Code.

Author: ```suggestion
ESLint is a tool for checking the quality and style of your client TypeScript
and template HTML code, and can provide valuable "live" feedback on
your coding in VS Code.
```

Since this is the first time we will be using ESLint there is an additional step to make sure the VS Code extension is working in the project. When you first open a TypeScript file you will see at the bottom right that ESLint is disabled.

![image](https://user-images.githubusercontent.com/1300395/107999308-bc59ec80-6fac-11eb-9784-75a471a50aa4.png)

Click the red "ESLINT" to open this dialog:

![image](https://user-images.githubusercontent.com/1300395/107996971-528b1400-6fa7-11eb-89bc-afc71747f820.png)

Click "Allow Everywhere" to enable ESLint.

You can also open this dialog with the following steps:

1. Hit `CTRL + SHIFT + P` (`⌘ + ⇧ + P` on Macs) to open the Command Palette. You can also find this by going to the "View" menu and clicking "Command Palette..."
2. Start typing and select "ESlint: Manage Library Execution". That should open a dialog seen above.

### Seeding the Database

> To give yourself some data to work with instead of starting with an empty database in our development environment, you need to 'seed' the database with some starter data. Seed data and the seed script are stored in the top level directory `database`. To seed the database, move into that directory and run `./mongoseed.sh` (or `.\mongoseed.bat` on Windows). This will take each of the JSON files in `database/seed/` and insert their elements into the `dev` database. (If you want to
> seed a different database, you can provide the name of the desired
> database as an argument to the `mongoseed` script.).

Author: ```suggestion
To give yourself some data to work with instead of starting with an empty database in our development environment, you need to 'seed' the database with some starter data. Seed data and the seed script are stored in the top level directory `database`. To seed the database, move into that directory and run `./mongoseed.sh` (or `.\mongoseed.bat` on Windows). This will take each of the JSON files in `database/seed/` and insert their elements into the `dev` database.
```

From the `client` directory:

- `ng test` runs the client tests
  - This will pop up a Chrome window with the results of the tests.
  - This will run "forever", updating both in your terminal and in the Chrome
    window that gets generated. Typing CTRL-C in the terminal window will end
    the `ng test` process and close the generated Chrome window.
  - You can add `ng test --watch=false` if you just want to run the tests once
    instead of going into the "run forever" mode.
- `ng test --code-coverage` runs the client tests and generates a coverage report
  - It generates a coverage report you can find in your client directory `client/coverage/client/index.html`.
  - Right click on `index.html` and select `Copy path` and paste it into your browser of choice. You can also drag and drop `index.html` onto the tab area of your browser and it will open it.

#### Linting the client

> We have included a tool called ESLint which helps analyze the client
> TypeScript and JavaScript code and catch various errors and concerns. You will most likely see it directly in VS Code as yellow and red underlines. You can also directly run the linter on the entire client by running `ng lint` in the terminal in the `client` directory. This will check the whole client project and tell you if there are any issues.

Author: ```suggestion
We have included a tool called ESLint which helps analyze the client
TypeScript and template HTML code and catch various errors and concerns. You will most likely see it directly in VS Code as yellow and red underlines. You can also directly run the linter on the entire client by running `ng lint` in the terminal in the `client` directory. This will check the whole client project and tell you if there are any issues.
```

  - The `title` element

## Deployment

Instructions on how to crate a DigitalOcean Droplet and setup your project are in [DEPLOYMENT.md](DEPLOYMENT.md).

## Resources

### Angular (client)

* [Angular Unit Testing (Karma)](https://angular.io/guide/testing)
* [Angular Routing](https://angular.io/guide/router)
* [Angular Forms](https://angular.io/guide/forms-overview)
* [Angular Material](https://material.angular.io/)
* [What are environments in Angular](https://angular.io/guide/build#configuring-application-environments)
* [Angular CLI](https://angular.io/cli)

### Javalin (server)

- [Javalin Documentation](https://javalin.io/documentation)
- [Javalin Tutorials](https://javalin.io/tutorials/)
  - [Testing Javalin Applications](https://javalin.io/tutorials/testing)
  - [Mocking Javalin classes in Mockito](https://javalin.io/tutorials/mockito-testing)

### MongoDB (database)

- [The MongoDB Manual](https://docs.mongodb.com/manual/)
- [MongoDB Java Drivers](https://mongodb.github.io/mongo-java-driver/)
  - [MongoDB Driver 3.12 Documentation](https://mongodb.github.io/mongo-java-driver/3.12/driver/)
- [MongoJack](https://mongojack.org/)

> - [JSON generator](https://next.json-generator.com/) for generating

Author: ```suggestion
- [JSON Generator](https://next.json-generator.com/) for generating
```
